File: 20th_march/2nd_second.cpp
#include <bits/stdc++.h>
using namespace std;

struct Edge {
    int u, v, w;
};

bool cmp(Edge a, Edge b) {
    return a.w < b.w;
}

vector<int> parent;

int find(int x) {
    if (parent[x] == x) return x;
    return parent[x] = find(parent[x]);
}

void unite(int a, int b) {
    a = find(a);
    b = find(b);
    if (a != b) parent[a] = b;
}

int kruskal(int n, vector<Edge>& edges, int skipEdgeIndex) {
    parent.resize(n + 1);
    for (int i = 1; i <= n; i++) parent[i] = i;

    int weight = 0, count = 0;

    for (int i = 0; i < edges.size(); i++) {
        if (i == skipEdgeIndex) continue;

        int u = edges[i].u;
        int v = edges[i].v;

        if (find(u) != find(v)) {
            unite(u, v);
            weight += edges[i].w;
            count++;
        }
    }

    if (count == n - 1) return weight;
    return INT_MAX;
}

int main() {
    int n, m;
    cin >> n >> m;

    vector<Edge> edges(m);

    for (int i = 0; i < m; i++) {
        cin >> edges[i].u >> edges[i].v >> edges[i].w;
    }

    sort(edges.begin(), edges.end(), cmp);

    // Step 1: Find MST
    int mst = kruskal(n, edges, -1);

    // Step 2: Find second MST
    int secondMST = INT_MAX;

    for (int i = 0; i < m; i++) {
        int val = kruskal(n, edges, i);
        if (val > mst) {
            secondMST = min(secondMST, val);
        }
    }

    cout << mst << " " << secondMST << endl;
}

File: 20th_march/2nd_second.py
class Edge:
    def __init__(self, u, v, w):
        self.u = u
        self.v = v
        self.w = w


# DSU (Disjoint Set)
def find(parent, x):
    if parent[x] != x:
        parent[x] = find(parent, parent[x])
    return parent[x]


def union(parent, a, b):
    a = find(parent, a)
    b = find(parent, b)
    if a != b:
        parent[a] = b


def kruskal(n, edges, skip_index):
    parent = [i for i in range(n + 1)]
    weight = 0
    count = 0

    for i in range(len(edges)):
        if i == skip_index:
            continue

        u, v, w = edges[i].u, edges[i].v, edges[i].w

        if find(parent, u) != find(parent, v):
            union(parent, u, v)
            weight += w
            count += 1

    if count == n - 1:
        return weight
    return float('inf')


# Main
n, m = map(int, input().split())

edges = []
for _ in range(m):
    u, v, w = map(int, input().split())
    edges.append(Edge(u, v, w))

# Sort edges by weight
edges.sort(key=lambda e: e.w)

# Step 1: MST
mst = kruskal(n, edges, -1)

# Step 2: Second MST
second_mst = float('inf')

for i in range(m):
    val = kruskal(n, edges, i)
    if mst < val < second_mst:
        second_mst = val

print(mst, second_mst)